### ISSUE TITLE
Option to unfreeze encoder layers when training a computer vision model

### ISSUE THREAD

**Issue by saad-palapa** [ludwig-ai/ludwig#3733]
**Is your feature request related to a problem? Please describe.**
Our computer vision solution is missing an important feature to improve accuracy. 

**Describe the use case**
In the past, I've followed this [Tensorflow tutorial](https://www.tensorflow.org/tutorials/images/transfer_learning#fine_tunin) to train an image classifier. This is a transfer learning technique that has two training rounds:
1. **Training:** add a classification head to a pretrained base model. The base model is frozen and the classification layers are trainable. This is currently what we have right now.
2. **Fine-tuning:** unfreezing layers from the base model and doing [10 epochs of low learning rate training](https://www.tensorflow.org/tutorials/images/transfer_learning#compile_the_model_2).

**Describe the solution you'd like**
There are two ways to implement this:
1. Update the configuration to let users choose to automatically run the 2nd fine-tuning round. Maybe this can be the default.
2. Rely on users to code their own fine-tuning round.

Both require a way to choose how to freeze layers. There are two approaches to freezing layers:
1. By percentage or layer count of the base model starting from the first hidden layer
2. Using a regex to choose which layers to freeze. [From example](https://github.com/google/automl/blob/c7392f2bab3165244d1c565b66409fa11fa82367/efficientdet/tf2/README.md#8-finetune-on-pascal-voc-2012-with-detector-coco-ckpt):
```
var_freeze_expr: '(efficientnet|fpn_cells|resample_p6)'
```

The later layers encode high level features of the image while the early ones encode low level features (corners, edges, colors, gradients, etc). It's not good to fine-tune the lower levels because the model will overfit those features.

When fine-tuning, we unfreeze the later layers of the pretrained base model. The number of layers to unfreeze depends on the task similarity and dataset size relative to the pretrained encoder.

![image](https://github.com/ludwig-ai/ludwig/assets/77132964/e2150c55-96af-4dc6-bf81-b09e96cd10ba)

**Describe alternatives you've considered**
The alternative is to keep things as they are and rely on users to implement their own 2-round transfer-learning job.
